### PR TITLE
interfaces: give priority to desktop-launch over desktop-legacy

### PR DESCRIPTION
**Pull request by sergio-costas**
The interface 'desktop-legacy' (and 'unity7') specifically denies read access to the .desktop files, which means that any extension that requires it (like gnome or kde) won't be able to read them.

Unfortunately, there are some specific cases where reading the .desktop files is mandatory, like when implementing the new Refresh Awareness specification. This specification requires to show the "visible name" of a snap, and its icon, and in order to have access to that, it is mandatory to be able to read the .desktop files.

The 'desktop-launch' interface does include read access to the .desktop files. Although it is a very privileged interface, it is not a problem because the snaps that implement the Refresh Awareness specification are too, so using it to gain access to the .desktop files should be enough. Unfortunately, mixing it with 'desktop-legacy' interface (which happens when the snap implementing the Refresh Awareness specification also uses the gnome or the kde extension) results in not having access to the files, because the 'deny' rules set by the later have priority over any 'allow' rule set by the former.

This PR adds a check when adding the specific .desktop rules in the 'desktop-legacy' interface: if the snap has a plug for the 'desktop-launch' interface, it won't apply the .desktop rules. This is not a problem, because without them, no access is granted by default (the rules added by 'desktop-legacy' allow to list the .desktop files, but not read them).

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

### CODE REVIEW

**Comment by sergio-costas** [comment]
Another possibility would be to create an specific interface to allow to read .desktop files, and give it priority instead of "desktop-launch".

File: interfaces/builtin/utils.go
Comment (zyga) [inline interfaces/builtin/utils.go:118]: This will check the name of the plug, not the interface of the plug. The name should not carry semantics.

**Review by zyga — CHANGES_REQUESTED** [review]
I have no opinion on the complexity of combining interfaces but the implementation needs a tweak.

File: interfaces/builtin/utils.go
Comment (sergio-costas) [reply interfaces/builtin/utils.go]: Good point! I think that I fixed it. Can you review it again when you have some spare time, please?

**Review by zyga — COMMENTED** [review]
The change looks good. I don't have opinion on the extra complexity.

**Review by zyga — APPROVED** [review]

File: interfaces/builtin/utils.go
Comment (sergio-costas) [reply interfaces/builtin/utils.go:126]: No, it shouldn't. It just checks if the snap uses the "desktop-launch" interface, but not if it is connected or not. As I comment in the description, it's not a security problem because, without these rules, no access is granted to the `XXXXX/applications` folders (while with these rules, you can read the list of files, but can't read the contents of the files), so a snap that includes the "desktop-launch" interface but doesn't connect it will have less rights than a snap that doesn't include it.

**Comment by sergio-costas** [comment]
The big problem is that using the _gnome_ extension forces the use of desktop-legacy, which forces being able to read the list of files in the `applications` folder, but not the data in those files. It seems to be done to avoid noise in the journal when XDG libraries try to read the .desktop files, according to `interfaces/builtin/utils.go`->`getDesktopFileRules()'.

**Review by alexmurray — APPROVED** [review]
LGTM for the stated purpose from a security point-of-view - although I agree the complexity seems a bit high and I don't love the idea of essentially hard-coding the 'desktop-launch' interface within utils.go - but this feels more like a question for @pedronis etc.

**Comment by sergio-costas** [comment]
I have to say that, yes, this approach is a bit "unorthodox" and "ugly", so maybe another way could be to add an specific API call, allowed only if the `desktop-launch` interface is connected, that allows to read .desktop files if the path is known... I can do a PR proposal... or maybe a document proposal... what do you think?

File: interfaces/builtin/utils.go
Comment (pedronis) [inline interfaces/builtin/utils.go:121]: this doesn't quite mean that the interface is actually connected

Comment (pedronis) [reply interfaces/builtin/utils.go:121]: the cleaner solution would be to think through some sort of optional conflict mechanism for snippets coming from different interfaces

**Review by pedronis — COMMENTED** [review]
some comments, I'm slightly worried about the maintainability of this approach

File: interfaces/builtin/utils.go
Comment (sergio-costas) [reply interfaces/builtin/utils.go:121]: I know that it doesn't mean that the interface is actually connected, but in this case that's not a problem because the snippet added just gives more privileges (allows to read the list of files at the DESKTOP folders), so if it isn't added because the other interface is requested, then the snap just will have less privileges because it won't be able to read anything from those folders.

Anyway, I agree in that it's an ugly way of doing things. About how to think about the conflicting mechanism... the problem that I see is that doing it very generic would require a parser for apparmor, which clearly is something excessive. The only idea that I think that could be implemented in that way is to identify blocks of snippets with some kind of unique ID, and that each interface that tries to implement that block will specify a priority, thus only the block with the highest priority would be finally added.

The other solution could be to just implement a REST API call that allows to read the contents of a .desktop file...

Comment (alexmurray) [reply interfaces/builtin/utils.go:121]: FWIW as part of new development work AppArmor is adding support for rule priorities itself - so this could eventually be used as a solution here but currently it is not available so perhaps in snapd we need to encode priorities for snippets as suggested. (see https://docs.google.com/presentation/d/13KMOgK7NPvWO51WdwEjik2HANTTU_drdjID7CwrQkdQ/edit#slide=id.g2c72fb4f079_0_56 as an example) (cc @jrjohansen)

Comment (jrjohansen) [reply interfaces/builtin/utils.go:121]: base priorities are ready now. There could be a few more tests. What is missing is
- logprof/genprof support, which isn't needed for snaps
- priority blocks, which might be nice to have for snippets. It depends on if you want to apply priority to the a whole block of rules in one go.

The priority blocks are something that might happen this week, as that is code that is getting shared with other important work.

Comment (sergio-costas) [reply interfaces/builtin/utils.go:121]: Mmm... but is that compatible with old kernels? If using priorities break compatibility with old versions... (or am I missing something?)

Comment (jrjohansen) [reply interfaces/builtin/utils.go:121]: It is compatible with old kernels. The compiler resolves the priorities as part of the compile.

Comment (sergio-costas) [reply interfaces/builtin/utils.go:121]: Oh! That makes sense. Thanks!

**Review by pedronis — CHANGES_REQUESTED** [review]
I'm ok to land this on a temporary basis as long as we add also a less implementation-details level test

File: interfaces/builtin/utils.go
Comment (sergio-costas) [reply interfaces/builtin/utils.go:122]: I'll prepare that test.

About the future, what do you think about this approach? Let's say that I implement a method like this in the `spec` class:

    func (*spec Specification) AddSnippetWithPriority(snippet string, ID string, priority int) {
        ...
    }

that allows to add a snippet but specifying a priority. That way, both interfaces would add their snippet using the same ID, but desktop-launch would set a bigger priority than desktop-legacy. That function would check if there is already a snippet with the same ID, and in that case, if the priority is lower than the new one, in which case that snippet will be replaced.

I see two ways of keeping track of the snippets:

* the "quick and ugly one": add two comments with a specific format to specify the beginning and the ending of a snippet with priority, and containing the ID and the current priority. This would allow to replace the text in the current snippet.

* the "elegant but not-so-quick one": add a second dictionary that contains the snippets with priority, and modify all the methods that read the whole snippets to, instead, call a private method that gets the "normal" snippets and the prioritized ones and mix them in a single snippet.

What do you think?

Comment (sergio-costas) [reply interfaces/builtin/utils.go:122]: Added the extra test.

File: interfaces/builtin/desktop_launch_test.go
Comment (pedronis) [inline interfaces/builtin/desktop_launch_test.go:127]: I think this could use Not(testutil.Contains) ?

File: interfaces/builtin/utils.go
Comment (pedronis) [reply interfaces/builtin/utils.go:122]: thanks, for the test.  Yes, what you outline there is probably what we want to do up to some naming discussions, but definitely using the some variant of the "elegant but not-so-quick one" approach, that's why I'm ok landing this first

**Review by pedronis — APPROVED** [review]
thanks, as I said I'm ok with this on a temporary basis

**Comment by pedronis** [comment]
marking blocked just because I'm asking previous reviewers to review the new test

File: interfaces/builtin/desktop_launch_test.go
Comment (sergio-costas) [reply interfaces/builtin/desktop_launch_test.go]: Yesss!!! That is what I was trying to do. Changed. Thanks!

File: interfaces/builtin/utils.go
Comment (sergio-costas) [reply interfaces/builtin/utils.go:122]: Perfect. Then I'll prepare a little braindump document, to begin to discuss that. Thanks!

File: interfaces/builtin/desktop_launch_test.go
Comment (ZeyadYasser) [inline interfaces/builtin/desktop_launch_test.go:116]: desktop-launch slot is already connected in `desktopLaunchSuite.SetUpTest`, and can be accessed through `s.slot`. this line can be removed.

File: interfaces/builtin/desktop_launch_test.go
Comment (ZeyadYasser) [inline interfaces/builtin/desktop_launch_test.go:122]:
```suggestion
	err = apparmorSpec.AddConnectedPlug(builtin.MustInterface("desktop-legacy"), plug, s.slot)
```

**Review by ZeyadYasser — CHANGES_REQUESTED** [review]
Looks good, Thank you!

Small comment regarding tests.

File: interfaces/builtin/desktop_launch_test.go
Comment (sergio-costas) [reply interfaces/builtin/desktop_launch_test.go]: Good point both changes. Done.

**Review by ZeyadYasser — APPROVED** [review]
Thank you!

File: interfaces/builtin/utils.go
Comment (zyga) [inline interfaces/builtin/utils.go:122]: So I've been looking at this and I feel a bit stupid as my memory of this area is somewhat rusty but let me spell out my concern.

Is there anything changed, when `desktop-lauch` is disconnected, and we call `getDesktopFileRules` from `AppArmorConnectedPlug` in unity7.go?

We are looking at `*snap.Info`'s `Plugs` attribute which knows nothing about connections.

If we are happy with this specific semantics then can we please add a comment with a big FIXME to make clear this is going on before merging this?

CC @pedronis

**Review by zyga — COMMENTED** [review]
This looks _almost_ good. On a closer look I have one concern inline.

I'd love to be involved in any follow-up to make this sort of interaction more generic.

File: interfaces/builtin/utils.go
Comment (sergio-costas) [reply interfaces/builtin/utils.go:122]: No, nothing changes. It only checks if the snap also requests the plug, but not if it's already connected or not. It's not a problem because without the plug connected, the net result in this specific case is less privileges than with the plug connected, AND less privileges than without the plug declared, thus any snap that requests `desktop-launch` interface will have less privileges than the same snap without that requests.

Of course, this is an ugly hack; that's why I'm already working on a more generic mechanism to replace this code.

I'll add the FIXME comment.

Comment (sergio-costas) [reply interfaces/builtin/utils.go:125]: Added.

**Comment by ernestl** [comment]
Got go-ahead from @sergio-costas to merge

**Pull request merged**
